3_ecosystem/3_1_testing/tests/cucumber.rs

```diff
@@ -0,0 +1,192 @@
+            Ok(_) => Ok(Some(line)),
+            Err(_) => Err("cannot read from pipe".into()),
+        }
+    }
+
+    fn write_stdin_line<S: AsRef<str>>(&mut self, s: S) -> Result<(), Error> {
+        event!(Level::INFO, "Write to a child thread");
+        self.stdin
+            .write_str(format!("{}\n", s.as_ref()))
+            .map_err(|_| "cannot write to child".into())
+    }
+}
+
+impl Drop for GameWorld {
+    fn drop(&mut self) {
+        let _ = self.game.kill();
+    }
+}
+
+impl Default for GameWorld {
+    fn default() -> Self {
+        use std::process::Command;
+
+        let program = env!("CARGO_BIN_EXE_step_3_1");
+
+        let the_number: usize = rand::random::<usize>() % (MAX_NUMBER_TO_GUESS + 1);
+
+        let mut child = Command::new(program)
+            .arg(the_number.to_string())
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()
+            .expect("failed to spawn a process");
+
+        let pid = child.id();
+
+        let stdin = child.stdin.take().expect("cannot take stdin");
+        let stdout = BufReader::new(child.stdout.take().expect("cannot take stdout"));
+
+        Self {
+            game: child,
+            the_number,
+            pid,
+            stdin,
+            stdout,
+            //
+            strings_we_wrote: vec![],
+        }
+    }
+}
+
+// Steps are defined with `given`, `when` and `then` attributes.
+#[given("a program is running")]
+fn hb(world: &mut GameWorld) -> Result<(), std::borrow::Cow<'static, str>> {
+    match world.game.try_wait() {
+        Ok(Some(status)) => Err(format!("died: {status}").into()),
+        Ok(None) => Ok(()),
+        Err(e) => Err(format!("Error waiting: {e}").into()),
+    }
+}
+
+#[when(expr = "we pass string: {}")]
+fn pass(world: &mut GameWorld, what: String) -> Result<(), std::borrow::Cow<'static, str>> {
+    world.write_stdin_line(&what)?;
+    world.strings_we_wrote.push(what);
+    Ok(())
+}
+
+#[when("we pass winning number")]
+fn try_win(world: &mut GameWorld) -> Result<(), std::borrow::Cow<'static, str>> {
+    world.write_stdin_line(world.the_number.to_string())?;
+    world.strings_we_wrote.push(world.the_number.to_string());
+    Ok(())
+}
+
+// the structure of output
+
+// #Please input your guess.
+// #...
+// -#You guessed: {}
+// --# You win!
+// --# Too small!
+// --# Too big!
+
+#[then("program ignores a line")]
+fn ignores(world: &mut GameWorld) -> Result<(), std::borrow::Cow<'static, str>> {
+    let line = world.read_stdout_line()?;
+    assert_eq!(line, Some("Guess the number!\n".into()));
+
+    let line = world.read_stdout_line()?;
+    assert_eq!(line, Some("Please input your guess.\n".into()));
+
+    // let line = world.read_stdout_line()?;
+    // assert_eq!(line, Some("boo\n".into()));
+
+    let line = world.read_stdout_line()?;
+    assert_eq!(line, Some("Please input your guess.\n".into()));
+
+    Ok(())
+}
+
+#[then("program produces sane output")]
```

Comment:
У Вас тут очепятка. Такие моменты тоже нужно вычитывать перед ревью.

```suggestion
#[then("program produces same output")]
```

Reply:
Просто плохое имя, это тест на корректность при любых условиях - его можно в then везде воткнуть, правда наверно это плохая идея.